File: 6_tree/morris.go
package main

import "fmt"

// BinaryTreeNode 定义了一个泛型接口。
// 接口的类型参数 T 表示节点本身的具体类型，
// 必须包含获取左右子节点和设置右子节点的方法，
// 用于 Morris 遍历中的线程指针操作。
// 实现此接口的自定义节点类型可以是任意结构体。
//
// 例如：
//
//	type TreeNode struct {
//	    Val         int
//	    LeftNode    *TreeNode
//	    RightNode   *TreeNode
//	}
//	func (n *TreeNode) Left() *TreeNode      { return n.LeftNode }
//	func (n *TreeNode) Right() *TreeNode     { return n.RightNode }
//	func (n *TreeNode) SetRight(r *TreeNode) { n.RightNode = r }
type BinaryTreeNode[T any] interface {
	Left() T
	Right() T
	SetRight(T)
}

// MorrisInorder 在二叉树上执行中序遍历，
// 无需递归或显式栈，并在遍历时临时创建并移除“线程”指针以恢复树结构。
// T 必须同时实现 BinaryTreeNode[T] 接口和 comparable，以支持 nil 检测。
// visit 回调会被依次调用，参数类型即为遍历到的节点类型 T。
func MorrisInorder[T interface {
	BinaryTreeNode[T]
	comparable
}](root T, visit func(node T)) {
	cur := root
	var nilValue T
	for cur != nilValue {
		left := cur.Left()
		if left == nilValue {
			visit(cur)
			cur = cur.Right()
		} else {
			pred := left
			for {
				right := pred.Right()
				if right == nilValue || right == cur {
					break
				}
				pred = right
			}
			if pred.Right() == nilValue {
				pred.SetRight(cur)
				cur = cur.Left()
			} else {
				pred.SetRight(nilValue)
				visit(cur)
				cur = cur.Right()
			}
		}
	}
}

// TreeNode 定义了二叉搜索树节点
// 实现左右访问与 SetRight 方法
type TreeNode struct {
	Val       int
	LeftNode  *TreeNode
	RightNode *TreeNode
}

func (n *TreeNode) Left() *TreeNode      { return n.LeftNode }
func (n *TreeNode) Right() *TreeNode     { return n.RightNode }
func (n *TreeNode) SetRight(r *TreeNode) { n.RightNode = r }

// recoverTree 在 BST 中恢复正被交换的两个节点值
func recoverTree(root *TreeNode) {
	var first, second, prev *TreeNode
	// 中序遍历，寻找逆序对
	MorrisInorder[*TreeNode](root, func(node *TreeNode) {
		if prev != nil && prev.Val > node.Val {
			if first == nil {
				first = prev
			}
			second = node
		}
		prev = node
	})
	// 交换找到的两个节点的值
	if first != nil && second != nil {
		first.Val, second.Val = second.Val, first.Val
	}
}

// 示例测试
func main() {
	// 构造示例：交换了两个节点
	//   1
	//  / \
	// 3   4
	//  \  /
	//   2
	root := &TreeNode{Val: 1}
	root.LeftNode = &TreeNode{Val: 3}
	root.RightNode = &TreeNode{Val: 4}
	root.LeftNode.RightNode = &TreeNode{Val: 2}

	fmt.Println("恢复前中序：")
	MorrisInorder[*TreeNode](root, func(n *TreeNode) { fmt.Printf("%d ", n.Val) })
	fmt.Println()

	recoverTree(root)

	fmt.Println("恢复后中序：")
	MorrisInorder[*TreeNode](root, func(n *TreeNode) { fmt.Printf("%d ", n.Val) })
	fmt.Println()
}

File: 6_tree/morris.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def morris_inorder(root):
    """
    Morris 中序遍历生成器，无需额外栈或递归，
    逐个 yield 出访问到的节点（按中序顺序）。
    完成后，能保证树结构被恢复。
    """
    curr = root
    while curr:
        if curr.left:
            # 找到左子树的最右节点
            pred = curr.left
            while pred.right and pred.right is not curr:
                pred = pred.right
            if pred.right is None:
                # 第一次访问：建立线索，进入左子树
                pred.right = curr
                curr = curr.left
            else:
                # 第二次访问：移除线索，访问 curr，然后进入右子树
                pred.right = None
                yield curr
                curr = curr.right
        else:
            # 无左子树，直接访问，再进入右子树
            yield curr
            curr = curr.right


if __name__ == "__main__":
    INF = int(1e18)

    class Solution:
        def recoverTree(self, root: TreeNode) -> None:
            """
            在常数额外空间下，恢复仅有两个节点被交换的 BST。
            """
            first = second = None
            prev = TreeNode(-INF)

            for node in morris_inorder(root):
                if prev.val > node.val:
                    if first is None:
                        first = prev
                    second = node
                prev = node

            first.val, second.val = second.val, first.val  # type: ignore
